raft/c: fix an indefinite hang in transfer leadership

bharathv opened the pull request:
This PR fixes a timeout in raft leadership transfer request. The timeout is caused by a race condition in raft that results in  a stuck recovery_stm, even after losing leadership. Exact sequence of events below:

- Request to transfer leadership in term - 10
- Starts a recovery_stm to catchup a follower  term - 10
- recovery_stm detects the follower has already been dispatched the latest log index, blocks on a CV until it hears back from the follower (indefinite wait)
- A racy step_down from a failed replicate request (stm initiated)
- A new leader takes over and local term incremented - term - 11
- recovery_stm  (in term 10) is stuck waiting for the CV is signaled (which happens if the node assumes leadership again in the future or at shutdown, an indefinite hang if it never assumes leadership)

This PR signals the CV upon step down to unblock the recovery_stm which then detects it is has to shutdown due to a step down (loss of leadership) and this automatically unblocks the transfer leadership request which was waiting for it.

An unrelated (but deeper) issue this PR avoids fixing is why the step down happens when a transfer is already in progress. In this case it was initiated by an STM invariant that requires a step down upon first failure (required for correctness). One can argue that it is not the right behavior to step down when a transfer is already in progress. The fix for it is likely invasive and has correctness implications at the state machine level (eg: idempotency), so avoiding for now.


<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v24.3.x
- [x] v24.2.x
- [x] v24.1.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bharathv committed:
raft/recovery: log line with more context

bharathv commented:
/ci-repeat 1
skip-units
release
dt-repeat 20
tests/rptest/chaos_tests/single_fault_test.py::TxSubscribeTest

bharathv commented:
/ci-repeat 1
skip-units
release
dt-repeat 20
tests/rptest/chaos_tests/single_fault_test.py::TxSubscribeTest

bharathv commented:
/ci-repeat 1
skip-units
release
dt-repeat=20
tests/rptest/chaos_tests/single_fault_test.py::TxSubscribeTest

bharathv commented:
/ci-repeat 1
skip-units
release
dt-repeat=40
tests/rptest/chaos_tests/single_fault_test.py::TxSubscribeTest

vbotbuildovich commented:
<!--RETRIED FAILURES-->
<!--6e743142-c753-414a-b6a6-85ec3d9e5af1-->
<!--RETRIED FAILURES-->
<!--ce690a13-cd92-4d08-a2b8-2ac0cb873ba9-->
<!--RETRIED FAILURES-->
<!--5bbf51c8-3a24-43fa-93bf-c330b11a6966-->
<!--RETRIED FAILURES-->
<!--c686c64f-db6f-418b-a623-a01fbe1e8e56-->
<!--RETRIED FAILURES-->
<!--8c8466b9-d2fc-45cb-a83e-4e5b8408917e-->
<!--RETRIED FAILURES-->
<!--1253d2e8-93f2-4e19-97d3-50fc0dc9acd2-->
<!--RETRIED FAILURES-->
<!--904deab9-061b-42a3-9f2d-37ff4038d67e-->
<!--RETRIED FAILURES-->
<!--575e874f-a935-42d5-93dc-b351264229b2-->
<!--RETRIED FAILURES-->
<!--f3591338-e106-4b02-8591-0e79f7562116-->
<!--RETRIED FAILURES-->
<!--38e8558e-7606-45ae-b544-62145dff61a7-->
<!--RETRIED FAILURES-->
<!--1cb16239-c707-419a-bba7-d7dc6cc06add-->
<!--RETRIED FAILURES-->
<!--9dab1de9-79d0-48f3-8c61-6e3aa91653ff-->
<!--RETRIED FAILURES-->
<!--d64af0c7-7b41-4500-99ed-ace3fc4f14d3-->
<!--RETRIED FAILURES-->
<!--d3b0661d-e561-4604-a78b-79e932e457d1-->
<!--body-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59180#01938da2-2f44-4119-b339-a9f789515d7b
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59208#0193900d-3629-498d-9926-1981d1ebc9fc
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59208#0193900d-3627-405b-bd38-a5dc660bce74
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59208#0193900d-362c-4a1e-ac0e-951f86073cd9
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59208#0193900d-362e-4948-ba03-87047c5cbc15
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59208#01939012-94b1-4626-870d-7023ed923a22
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59208#01939012-94b3-4515-9a2c-ea0f2873d859
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59208#01939012-94b6-4e21-bd81-f895f8d48378
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59208#01939012-94b4-45b9-ad08-215f88af399b
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59270#019396fe-eff8-45ef-930c-938ccdfcd941
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59270#01939703-4a8e-4bfc-b5df-cff6c67541ce
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59270#01939703-4a91-43c6-8772-28491b120aa8
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59285#019397d8-7c8b-4ccf-9e84-6ffce3c607a6
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59285#019397d8-7c8e-48be-aadb-2c2c4c267a5e
<!--/body-->

bharathv committed:
raft/c: notify waiters on follower state reset

This typically happens when there is a stepdown and the downstream
consumers like recovery need to know about it.

bharathv commented:
/ci-repeat 1
skip-units
release
dt-repeat=40
tests/rptest/chaos_tests/single_fault_test.py::TxSubscribeTest

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59208-->
<!--956d1501-0a5e-44ef-82ab-7f8f5ba59aa6-->
<!--d592f80f-b16d-41bb-ab3f-8a6e8aa93a2e-->
<!--03ca78a5-a271-4356-8fdc-9b64c7c85f1c-->
<!--7ee0d8e3-1937-42fb-a3d3-e220f8b16ee6-->
<!--0e99ca23-84e3-4df6-a851-d6cd465c3d1d-->
<!--8cc787f8-5bbf-4f0e-8a6f-fd538342bf0d-->
<!--26bb37fc-92af-4267-a373-ec9596d62094-->
<!--0709162a-21c4-41df-8da0-5ead17eff03d-->
<!--title-->
## Retry command for [Build#59208](https://buildkite.com/redpanda/redpanda/builds/59208)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/license_enforcement_test.py::LicenseEnforcementTest.test_license_enforcement@{"clean_node_after_recovery":true,"clean_node_before_recovery":false}
tests/rptest/tests/license_enforcement_test.py::LicenseEnforcementTest.test_license_enforcement@{"clean_node_after_recovery":false,"clean_node_before_recovery":false}
tests/rptest/tests/license_enforcement_test.py::LicenseEnforcementTest.test_escape_hatch_license_variable@{"clean_node_before_upgrade":false}
tests/rptest/tests/license_enforcement_test.py::LicenseEnforcementTest.test_license_enforcement@{"clean_node_after_recovery":false,"clean_node_before_recovery":true}
tests/rptest/tests/license_enforcement_test.py::LicenseEnforcementTest.test_license_enforcement@{"clean_node_after_recovery":true,"clean_node_before_recovery":true}
tests/rptest/tests/license_enforcement_test.py::LicenseEnforcementTest.test_escape_hatch_license_variable@{"clean_node_before_upgrade":true}
```
<!--/body-->

vbotbuildovich commented:
<!--NEW FAILURES-->
<!--1a64c861-f2b3-4d75-9cd7-80e6bed9d6e9-->
<!--NEW FAILURES-->
<!--3f2433d8-471e-4bdb-ac83-233de8a2a646-->
<!--NEW FAILURES-->
<!--02fcbe71-0704-4e52-9c5b-0bafb91ec35c-->
<!--NEW FAILURES-->
<!--0727b86a-cfaa-4910-8303-61204c9d8b50-->
<!--body-->
non flaky failures in https://buildkite.com/redpanda/redpanda/builds/59208#0193900d-362c-4a1e-ac0e-951f86073cd9:
```
"rptest.tests.license_enforcement_test.LicenseEnforcementTest.test_license_enforcement.clean_node_before_recovery=True.clean_node_after_recovery=False"
```
non flaky failures in https://buildkite.com/redpanda/redpanda/builds/59208#0193900d-362e-4948-ba03-87047c5cbc15:
```
"rptest.tests.license_enforcement_test.LicenseEnforcementTest.test_escape_hatch_license_variable.clean_node_before_upgrade=True"
```
non flaky failures in https://buildkite.com/redpanda/redpanda/builds/59208#01939012-94b1-4626-870d-7023ed923a22:
```
"rptest.tests.license_enforcement_test.LicenseEnforcementTest.test_license_enforcement.clean_node_before_recovery=True.clean_node_after_recovery=False"
```
non flaky failures in https://buildkite.com/redpanda/redpanda/builds/59208#01939012-94b3-4515-9a2c-ea0f2873d859:
```
"rptest.tests.license_enforcement_test.LicenseEnforcementTest.test_escape_hatch_license_variable.clean_node_before_upgrade=True"
```
<!--/body-->

mmaslankaprv commented on src/v/raft/recovery_stm.cc:
```diff
@@ -708,7 +708,11 @@ ss::future<> recovery_stm::apply() {
-          vlog(_ctxlog.trace, "Finished recovery");
+          vlog(
+            _ctxlog.trace,
+            "Finished recovery of node: {} in term: {}",
```
nit: node id is not required as we already have it in: 
```
  , _ctxlog(
      raftlog,
      ssx::sformat(
        "[follower: {}] [group_id:{}, {}]",
        _node_id,
        _ptr->group(),
        _ptr->ntp()))
```

mmaslankaprv commented:
/ci-repeat 1

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
/backport v24.1.x
